### ISSUE TITLE
NA OFI: strange HG_Cancel() behaviour

### ISSUE THREAD

**Issue by yulujia** [mercury-hpc/mercury#308]
using mercury commit >= b65ccb3066435 I experience some strange behavior with HG_Cancel().

case 1:
1) start to service procs proc_a and proc_b, each creates a NA context, and a dedicated progress thread.
2) proc_a sends RPC to proc_b, sleeps for 5 seconds, call HG_Cancel() to cancel the RPC. HG_Cancel() fails with HG_CANCEL_ERROR
3) proc_a fails during HG_Context_destroy(), error message:

09/03-23:47:22.31 CaRT[26606/26606] hg ERR # NA -- Error -- mercury/src/na/na_ofi.c:4790
na_ofi_cancel(): fi_cancel unexpected recv failed, rc: -2(No such file or directory).
09/03-23:47:22.31 CaRT[26606/26606] hg ERR # HG -- Error -- mercury/src/mercury_core.c:991
hg_core_pending_list_cancel(): Could not cancel handle
09/03-23:47:22.31 CaRT[26606/26606] hg ERR # HG -- Error -- mercury/src/mercury_core.c:3836
HG_Core_context_destroy(): Cannot cancel list of pending entries
09/03-23:47:22.31 CaRT[26606/26606] hg ERR # HG -- Error -- mercury/src/mercury.c:1230
HG_Context_destroy(): Could not destroy HG core context

case 2:
1) start to service procs proc_a and proc_b, each creates a NA context, and a dedicated progress thread.
2) proc_a sends RPC to proc_b, sleeps for 5 seconds, call HG_Cancel() to cancel the RPC. HG_Cancel() fails with HG_CANCEL_ERROR
3) call HG_Destroy() multiple times until the RPC handle is destroyed.
4) Later on, the RPC completion callback is triggered anyway, and we get a segfault since the hg_handle is already destroyed.

**Comment by dmiter** [comment]
Mercury code assumes that the `fi_cancel` operation is blocking and will always actually cancel the operation. These assumptions are incorrect.

Here is the cancel text from the libfabric man page:

>fi_cancel
>fi_cancel attempts to cancel an outstanding asynchronous operation. Canceling an operation causes the fabric provider to search for the operation and, if it is still pending, complete it as having been canceled. An error queue entry will be available in the associated error queue with error code FI_ECANCELED. On the other hand, if the operation completed before the call to fi_cancel, then the completion status of that operation will be available in the associated completion queue. No specific entry related to fi_cancel itself will be posted.
>
>Cancel uses the context parameter associated with an operation to identify the request to cancel. Operations posted without a valid context parameter – either no context parameter is specified or the context value was ignored by the provider – cannot be canceled. If multiple outstanding operations match the context parameter, only one will be canceled. In this case, the operation which is canceled is provider specific. The cancel operation is asynchronous, but will complete within a bounded period of time.

.. and also ..

>RETURN VALUES
>Returns 0 on success. On error, a negative value corresponding to fabric errno is returned. For fi_cancel, a return value of 0 indicates that the cancel request was submitted for processing.

For example, this Mercury code immediately marks the Mercury request as "complete" as soon as the `fi_cancel` function returns.  Later, when the operation actually completes and is reported via `fi_cq_readfrom()`, Mercury reports an error because the "request is already completed".

**Comment by soumagne** [comment]
@dmiter thanks for the additional details, I will get back to you once I look at the cancel issue and @liuxuezhao pull request, which is sort of related.

**Comment by liuxuezhao** [comment]
hmm, the cancel operation is a kind of important and complex for some cases.

Based on some debugging experience long time ago (maybe in year 2017 ~2018 when I working on the OFI plugin), the OFI level cancel behavior is not fully same as what its man-page described. We met quite some troubles for some DAOS use cases that killing server process and want it recover from RPC failure/timeout.  
When I wrote the initial OFI plugin code, at very beginning I was follow what suggested from its man-page, that when fi_cancel returns success then don't complete it immediately, instead it will wait the FI_ECANCELED event reported by fi_cq_readerr(). But in the real test it make it hard to pass kinds of DAOS testing (most important is rebuild testing that with kinds of process killing test), at that time after some debugging that it looks like 1) got duplicated events from OFI sometimes, and/or 2) lost some events from OFI then cannot tigger the mercury level completion then cart level completion callback cannot be triggered.
Considering together with the situation that OFI level can-only cancel recv operation, cannot cancel send or inflight RMA read or write operation. Then I made some change to the cancel handling to be that right after fi_cancel() called we immediately complete it but ignore the possibly will (or will not) be reported  FI_ECANCELED event from OFI later. So with the ugly/tricky na_ofi_op_id_valid() related verification/checking. 
That probably is not the best approach, but it make DAOS pass kinds of testing with process killing at different time. 

But mercury level cancel handing with some other changes later and actually DAOS cannot work very well for some new versions, the latest workable version in my testing is with c68870ffc0409c29eece5ba036c6efd3c22cee41. But I did not test with latest version.

That is just to explain a little bit about the cancel handling. I think that it will be great if we could have some special test cases at mercury level that with server killing at different time ( for example killing server before client do addr_lookup, or after addr_lookup but before send RPC, or after send RPC before server reply, then after some time no reply got by client it do HG_Cancel() to see if its completion cb can be triggered with appropriate hg_cb_info::ret set, or the HG_Forward() returns error for some cases.
@yulujia @dmiter we probably can also add some special testing cases for that at cart level.

Probably the most correct way is handling it just as ofi man-page described, and if it does work then fix it at OFI level. But sometimes it is hard to reproduce it at ofi level and provide fixing for some special cases.
For cancel related changes it looks better to verify it with above test cases, before those test cases added we probably can test it with daos test (rebuild test is most important).
